### ISSUE TITLE
egress MTLS origination from sidecar

### ISSUE THREAD

**Issue by arjun-beathi** [istio/istio#45259]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

I followed this documentation for initiating MTLS connection directly from sidecar ( instead of egress gateway)
https://istio.io/latest/docs/tasks/traffic-management/egress/egress-tls-origination/#configure-mutual-tls-origination-for-egress-traffic-at-sidecar

I already have external service which hosted as mtls.

Calls are going to external in plain text instead of negioting mtls.

apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: abc-https
  namespace: abc-namespace
spec:
  hosts:
  - abc-mtls.com
  location: MESH_EXTERNAL
  ports:
  - name: https-port
    number: 443
    protocol: HTTPS
  resolution: DNS
  
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: originate-mtls-api
  namespace: abc-namespace
spec:
  exportTo:
  - .
  workloadSelector:
    matchLabels:
      app: abc
  host: abc-mtls.com
  trafficPolicy:
    loadBalancer:
      simple: ROUND_ROBIN
    portLevelSettings:
    - port:
        number: 443
      tls:
        mode: MUTUAL
        credentialName: abc-credential
        sni: abc-mtls.com
        
 istio log:   
 {"istio.req.protocol":"HTTP/1.1","istio.req.id":"73418012-8cd1-470c-ac90-6ccbb3f0c9d5","istio.client.address":"100.67.128.155:58758","istio.resp.code":**308**,"istio.client.local.address":"xx.xxx.xxx.xx:80","istio.upstream.trasport.fail.reason":null,"istio.resp.flag":"-","istio.upstream.cluster":"outbound|443||abc-mtls.com","istio.req.method":"POST","istio.req.user.agent":"ReactorNetty/1.0.30","istio.upstream.host":"xx.xxx.xxx.xx:443","istio.duration":102,"istio.conn.termination.details":null,"istio.resp.upstream.time":"102","istio.req.start.time":"2023-06-02T19:34:23.253Z","istio.req.path":"/usb/pcd-services-data/graphql-data/v1/graphql","istio.req.host":"abc-mtls.com","istio.upstream.request.attempt.count":1,"istio.req.x-forward-for":null}

I also tried adding virtualservice to route any traffic on 80 port to 443, still it doesn't help.

apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: direct-abc-api-op-through-sidecar-proxy
  namespace: abc-namespace
spec:
  exportTo:
  - .
  gateways:
  - mesh
  hosts:
  - abc-mtls.com
  http:
  - match:
    - gateways:
      - mesh
      port: 80
    route:
    - destination:
        host: abc-mtls.com
        port:
          number: 443
      weight: 100
 
 
 From the main pod, we are making http call. curl -iv http://abc-mtls.com.
 
 Also, it works when I configure it through egress gateway. (https://istio.io/latest/docs/tasks/traffic-management/egress/egress-gateway-tls-origination/#configure-mutual-tls-origination-for-egress-traffic)

### Version

```Text
istioctl version
client version: 1.13.5
control plane version: 1.17.2
data plane version: 1.17.2 (13 proxies)

kubectl version --short
Client Version: v1.25.9
Server Version: v1.24.10
```

### Affected product area

- [ ] Ambient
- [X] Docs
- [ ] Installation
- [X] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

**Comment by hzxuzhonghu** [comment]
@kfaseela

**Comment by kfaseela** [comment]
@arjun-beathi could you please share the output of `istioctl proxy-config  secret` for the sidecar where the DR configuration is applied?

**Comment by arjun-beathi** [comment]
> @arjun-beathi could you please share the output of `istioctl proxy-config secret` for the sidecar where the DR configuration is applied?

`RESOURCE NAME                                       TYPE           STATUS     VALID CERT     SERIAL NUMBER                               NOT AFTER                NOT BEFORE
default                                             Cert Chain     ACTIVE     true           23931*******755025832909117     2023-06-06T06:55:59Z     2023-06-05T06:53:59Z
kubernetes://abc-credential            Cert Chain     ACTIVE     true           160***8050                                  2024-06-29T21:04:42Z     2023-06-01T20:34:42Z
ROOTCA                                              CA             ACTIVE     true           72073******026002558466188      2033-03-18T20:14:58Z     2023-03-21T20:14:58Z
kubernetes://abc-credential-cacert     CA             ACTIVE     true           133***4750                                  2030-11-30T18:15:07Z     2012-05-31T17:45:07Z`

**Comment by arjun-beathi** [comment]
Isn't  main pod makes http call then the sidecar will convert it to https/mtls call based on istio condifuration?

I tried both http and https curl call from main pod.

HTTPS call:
curl -iv https://abc-mtls.com
*   Trying xx.xxx.xxx.xx:443...
* Connected to iabc-mtlscom (xx.xxx.xxx.xx) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* error:1408F10B:SSL routines:ssl3_get_record:wrong version number
* Closing connection 0
curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number

HTTP call:
curl -iv http://abc-mtls.com
*   Trying xx.xxx.xxx.xx:80...
* Connected to abc-mtls.com (xx.xxx.xxx.xx) port 80 (#0)
> GET / HTTP/1.1
> Host: abc-mtls.com
> User-Agent: curl/7.80.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 308 Permanent Redirect
HTTP/1.1 308 Permanent Redirect
< date: Mon, 05 Jun 2023 15:34:56 GMT
date: Mon, 05 Jun 2023 15:34:56 GMT
< content-type: text/html
content-type: text/html
< content-length: 164
content-length: 164
< location: https://abc-mtlscom
location: https://abc-mtls.com
< strict-transport-security: max-age=15724800; includeSubDomains
strict-transport-security: max-age=15724800; includeSubDomains
< x-envoy-upstream-service-time: 217
x-envoy-upstream-service-time: 217
< server: envoy
server: envoy

<
<html>
<head><title>308 Permanent Redirect</title></head>
<body>
<center><h1>308 Permanent Redirect</h1></center>
<hr><center>nginx</center>
</body>
</html>


HTTP call is going upto the server, server redirects to https.

**Comment by kfaseela** [comment]
I agree that example used in the document should have been better. But as you can see the current curl command in the example is `kubectl exec "$(kubectl get pod -l app=sleep -o jsonpath={.items..metadata.name})" -c sleep -- curl -sS http://my-nginx.mesh-external.svc.cluster.local:443` . Let me know if that works.

**Comment by arjun-beathi** [comment]
hmm. it doesn't work with that as well. getting curl: (56) Recv failure: Connection reset by peer.

Can you clarify on serviceentry as well?

**Comment by kfaseela** [comment]
@arjun-beathi : there is no need for service entry if you are directly using port 443 in the curl command. If you have followed the exact same example, have only the configuration given in the example. (ie, no SE and VS like you have added in the description). But, if you want the same nginx example to work for port 80, you can use the SE and DR like below:

```
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: first-service-entry
spec:
  exportTo:
  - .
  hosts:
  - my-nginx.mesh-external.svc.cluster.local
  location: MESH_EXTERNAL
  ports:
  - number: 80
    name: http-port
    protocol: HTTP
    targetPort: 443
  - number: 443
    name: https-port
    protocol: HTTPS
  resolution: DNS
---
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: first-destination-rule
spec:
  workloadSelector:
    matchLabels:
      app: sleep
  exportTo:
  - .
  host: my-nginx.mesh-external.svc.cluster.local
  trafficPolicy:
    loadBalancer:
      simple: ROUND_ROBIN
    portLevelSettings:
    - port:
        number: 80
      tls:
        credentialName: client-credential
        mode: MUTUAL
        sni: my-nginx.mesh-external.svc.cluster.local
```

**Comment by kfaseela** [comment]
> Agree if it is an internal service, we don't need service entry. The example is calling an internal service and we don't need to add serviceentry.
> 
> What Im trying is calling an external hosted service(which is mtls). In my case, I need service entry.

you can use the above SE and DR config, and do a curl on port 80 itself. it will be internally redirected to 443. In your issue description you have configured your SE and DR only for port 443, and were trying to curl with port 80.

**Comment by arjun-beathi** [comment]
> > Agree if it is an internal service, we don't need service entry. The example is calling an internal service and we don't need to add serviceentry.
> > What Im trying is calling an external hosted service(which is mtls). In my case, I need service entry.
> 
> you can use the above SE and DR config, and do a curl on port 80 itself. it will be internally redirected to 443. In your issue description you have configured your SE and DR only for port 443, and were trying to curl with port 80.

Thanks @kfaseela for documentation updates!!

Yes, I tried this option before as well. Calls are reaching the external endpoint on port 443 in "plain" text and getting a redirect 308.

corresponding istio log in json:
{"istio.upstream.request.attempt.count":1,
"istio.req.user.agent":"curl/7.80.0",
"istio.req.id":"0020f0d4-d1f4-4854-8f21-bf3d4103f4f8",
"istio.resp.flag":"-",
"istio.upstream.cluster":"outbound|80||abc-mtls.com",
"istio.upstream.host":"10.100.200.20:443",
"istio.resp.upstream.time":"184",
"istio.req.path":"/abc/abc-services-data/graphql-data/v1/graphql",
"istio.req.host":"abc-mtls.com",
"istio.req.x-forward-for":null,
"istio.client.address":"100.67.128.110:50026",
"istio.upstream.trasport.fail.reason":null,
"istio.resp.code":**308**,
"istio.req.protocol":"HTTP/1.1",
"istio.req.method":"POST",
"istio.client.local.address":"10.100.200.20:80",
"istio.req.start.time":"2023-06-06T13:51:30.359Z",
"istio.conn.termination.details":null,
"istio.duration":185}

**Comment by kfaseela** [comment]
Please paste your complete configuration steps one by one, and the curl command also. Difficult to debug otherwise.

**Comment by arjun-beathi** [comment]
```
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: abc-mtls-domain-com-https
spec:
  hosts:
  - abc-mtls.domain.com
  location: MESH_EXTERNAL
  ports:
  - name: https-port
    number: 443
    protocol: HTTPS
  - name: http-port
    number: 80
    protocol: HTTP
    targetPort: 443
  resolution: DNS

apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: originate-mtls-for-abc
spec:
  host: abc-mtls.domain.com
  trafficPolicy:
    loadBalancer:
      simple: ROUND_ROBIN
    portLevelSettings:
    - port:
        number: 80
      tls:
        credentialName: abc-credential
        mode: MUTUAL
        sni: abc-mtls.domain.com
  workloadSelector:
    matchLabels:
      app: sleep
```
      
curl command: curl http://abc-mtls.domain.com
```
<html>
<head><title>308 Permanent Redirect</title></head>
<body>
<center><h1>308 Permanent Redirect</h1></center>
<hr><center>nginx</center>
</body>
</html>
```

envoy configuration on that proxy:
![image](https://github.com/istio/istio/assets/117765555/5a1aa729-7a1c-46af-918f-62e69705b57a)


I will ping in slack when I get a chance.

**Comment by arjun-beathi** [comment]
I tested with the example in the document, that works. I also did tests on target which is hosted on istio-ingressgateway, that works as well.
The only target which is not working is target hosted behind nginx controller (both simple and mutual tls). The same target work through egress gateway.

Now Im not sure whether it is istio issue or the target nginx.  ( if it is target, why calls are working through egress gateway)

Hope somebody can take a look at this.
